Add Folia support

jackh54 opened the pull request:
This PR adds Folia support. 

• Added Folia async scheduler support with automatic detection and fallback to Bukkit scheduler
• Fixed NullPointerException in Tasks by reordering initialization - Query is now created before Tasks (Ran into the same issue as #7 )
• Added null check in Tasks
• No breaking changes, backward compatible with existing servers
• Builds successfully and resolves the async task exception (gradle wrapper seems to have issues though, had to create a new one locally)

Tested on both Paper & Folia. Added this support due to [this PR](https://github.com/tebexio/Tebex-Minecraft/pull/84) which is to be released in the next version.

alexdev03 commented:
Hi, thanks for your pr. I'm unable to test that but that looks good except for a few things:
- I don't think there is need to use reflection, you could change the dependency to paperapi and you would have methods to call (Only if the platform is correct ofc)
- Instead of creating a BukkitRunnable for non folia platforms you could just call [this](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/scheduler/BukkitScheduler.html#runTaskTimerAsynchronously(org.bukkit.plugin.Plugin,java.lang.Runnable,long,long))
- The same for fallbackToLegacyScheduler

jackh54 commented:
Hi, thanks for the quick response. I have done the following:
- Switched to Paper API dependency
- Replaced BukkitRunnable with the link you provided
- Cleaned up the reflection code

Regarding reflection, I tried to get rid of it completely by importing paper classes, however it creates compile-time issues because those classes don't exist on regular bukkit/spigot servers.

I've tested on both a paper server and a folia server, with a Tebex account having 600+ payments. Seems to work fine, and all placeholders work.

alexdev03 commented:
Hi, thanks for the changes. I don't remember in what version paper added those scheduler methods but I think we could drop support for old versions that let use remove reflection parts. Nowdays more than 75% of servers are using 1.20+ (according to bstats)

jackh54 commented:
This should be good, tested and works well.

alexdev03 commented:
Thanks
